### PR TITLE
Migrated to null safety

### PR DESCRIPTION
**Pull request by abed-binary**
## Description of the change

Migrated the project to null safety.

### CODE REVIEW

File: lib/RudderProperty.dart [hunk 1]
```diff
@@ -1,7 +1,7 @@
 class RudderProperty {
-  Map<String, dynamic> __map = new Map();
+  Map<String?, dynamic> __map = new Map();
```
Comment (desusai7) [inline lib/RudderProperty.dart:2]:
```suggestion
 Map<String, dynamic> __map = new Map();
```

File: lib/RudderProperty.dart [hunk 2]
```diff
@@ -1,7 +1,7 @@
...
 
-  Map<String, dynamic> getMap() {
+  Map<String?, dynamic> getMap() {
```
Comment (desusai7) [inline lib/RudderProperty.dart:4]:
```suggestion
 Map<String, dynamic> getMap() {
```

File: lib/RudderProperty.dart [hunk 3]
```diff
@@ -14,15 +14,15 @@ class RudderProperty {
...
       this.__map.addAll(map);
       return this;
     }
     if (value is RudderProperty) {
       this.__map[key] = value.getMap();
     } else {
-      map[key] = value;
+      map![key] = value;
     }
     return this;
```
Comment (desusai7) [inline lib/RudderProperty.dart:27]: Can we please convert it into this way @abed-binary : 
```suggestion
  {String? key, dynamic value, Map<String, dynamic>? map}) {
    if (map != null) {
      this.__map.addAll(map);
      return this;
    }
    if (key != null) {
      if (value is RudderProperty) {
        this.__map[key] = value.getMap();
      } else {
        this.__map[key] = value;
      }
    }
    return this;
  }
```

**Comment by abed-binary** [comment]
I already accepted the CLA like 2 weeks ago. Please review the changes and let me know if everything is ok. thank you.

File: lib/RudderIntegration.dart [hunk 4]
```diff
@@ -1,4 +1,4 @@
 class RudderIntegration {
-  String getKey() {}
+  String? getKey() {}
```
Comment (desusai7) [inline lib/RudderIntegration.dart:2]: Can we please make the return type of `getKey()` Non-Nullable ?

File: lib/RudderOption.dart [hunk 5]
```diff
@@ -44,34 +46,30 @@ class RudderOption {
...
   RudderOption putIntegrationWithFactory(
       RudderIntegration factory, bool enabled) {
-    if (this.integrations == null) {
+    if (integrations == null) {
       integrations = new Map();
     }
-    if (factory != null && enabled != null) {
-      integrations[factory.getKey()] = enabled;
+    if (factory.getKey() != null) {
+      integrations[factory.getKey()!] = enabled;
```
Comment (desusai7) [inline lib/RudderOption.dart:62]: Since the `getKey()` doesn't return null. We can ignore the null check here

Comment (abed-binary) [reply lib/RudderOption.dart:62]: Screenshot 2021-06-03 at 12 10 51 PM

What do you advice to do ?

Comment (desusai7) [reply lib/RudderOption.dart:62]: Actually `RudderIntegration.dart` is used as a super class used for the Device mode integrations and not in the core SDK.
All the Device Mode integrations will implement `RudderIntegration.dart` and its mandatory for them to return 
the `getKey()` method in those classes to return a Non-Null value.
So I feel we can ignore this warning

File: lib/RudderOption.dart [hunk 6]
```diff
@@ -4,6 +4,7 @@ import './RudderIntegration.dart';
 // we left fetching the external ids from the scratch here
 class RudderOption {
   List<Map<String, dynamic>>? externalIds;
+  late Map<String, Object> integrations;
```
Comment (desusai7) [inline lib/RudderOption.dart:7]: `integrations` object can be nullable if we don't add any integrations. So please implement null-check for integrations object

File: lib/RudderOption.dart (hunk 5 above)
Comment (abed-binary) [reply lib/RudderOption.dart:62]: I see. Unfortunately, we can't ignore as it is an error not a warning, we have to return something or make it nullable to be able to build. So which should we go with ?

Comment (desusai7) [reply lib/RudderOption.dart:62]: Better we go by returning `"Integration"` from the `getKey()` in `RudderIntegration.dart`
This way we meet both the constraints.
WDYT ?

Comment (abed-binary) [reply lib/RudderOption.dart]: I returned `Integration` as suggested.

**Review by arnab-p — APPROVED** [review]

**Pull request merged**